server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -241,8 +241,18 @@ static NodeConstruction prepareConstruction(
     ) {
         List<Closeable> closeables = new ArrayList<>();
         try {
+            Settings envSettings = Settings.builder()
+                .put(initialEnvironment.settings())
+                .put(Client.CLIENT_TYPE_SETTING_S.getKey(), "node")
+                .build();
+
             NodeConstruction constructor = new NodeConstruction(closeables);
-            constructor.construct(initialEnvironment, serviceProvider, forbidPrivateIndexSettings);
+            Settings settings = constructor.createEnvironment(initialEnvironment, envSettings, serviceProvider);
```

Comment:
It's a good step and I see you are using returned types to enforce relationship and order between calls; however I find a bit confusing to have this function have both side effects (sets environment and pluginService) and return something. 
It's OK for now as it minimizes change and makes easier to see the PR does not change behaviour, but I would consider a better alternative for the final refactoring step.

Reply:
There's going to be lots of steps to this refactoring - this is only the first